src/nginx.py

```diff
@@ -0,0 +1,305 @@
+# Copyright 2023 Canonical
+# See LICENSE file for licensing details.
+"""Nginx workload."""
+
+import logging
+from typing import Dict, List, Set
+
+import crossplane
+from charms.mimir_coordinator_k8s.v0.mimir_cluster import MimirClusterProvider
+from ops.pebble import Layer
+
+logger = logging.getLogger(__name__)
+
+
+class Nginx:
+    """Helper class to manage the nginx workload."""
+
+    config_path = "/etc/nginx/nginx.conf"
+
+    def __init__(self, cluster_provider: MimirClusterProvider, *args):
+        super().__init__(*args)
+        self.cluster_provider = cluster_provider
+
+    @property
+    def config(self) -> str:
+        """Build and return the Nginx configuration."""
+        log_level = "error"
+        auth_enabled = False
+        addresses_by_role = self.cluster_provider.gather_addresses_by_role()
+
+        def upstreams(addresses_by_role: Dict[str, Set[str]]) -> List[Dict]:
+            nginx_upstreams = []
+            for role, address_set in addresses_by_role.items():
+                nginx_upstreams.append(
+                    {
+                        "directive": "upstream",
+                        "args": [role],
+                        "block": [
+                            {"directive": "server", "args": [f"{addr}:8080"]}
+                            for addr in address_set
+                        ],
+                    }
+                )
+
+            return nginx_upstreams
+
+        def log_verbose(verbose):
+            if verbose:
+                return [{"directive": "access_log", "args": ["/dev/stderr", "main"]}]
+            return [
+                {
+                    "directive": "map",
+                    "args": ["$status", "$loggable"],
+                    "block": [
+                        {"directive": "~^[23]", "args": ["0"]},
+                        {"directive": "default", "args": ["1"]},
+                    ],
+                },
+                {"directive": "access_log", "args": ["/dev/stderr"]},
+            ]
+
+        def resolver(custom_resolver):
+            if custom_resolver:
+                return [{"directive": "resolver", "args": [custom_resolver]}]
+            return [{"directive": "resolver", "args": ["kube-dns.kube-system.svc.cluster.local."]}]
+
+        def basic_auth(enabled):
+            if enabled:
+                return [
+                    {"directive": "auth_basic", "args": ['"Mimir"']},
+                    {
+                        "directive": "auth_basic_user_file",
+                        "args": ["/etc/nginx/secrets/.htpasswd"],
+                    },
+                ]
+            return []
+
+        # build the complete configuration
+        full_config = [
```

Comment:
Since this `full_config` is a very large config, shouldn't we have it in a separate file just for that?

Reply:
`src/nginx.py` **is** a separate file; all it does is building the Nginx configuration, and it needs to access the `cluster_provider` object (to get the workers' addresses), meaning we'd have to pass it yet another time, and `src/nginx.py` would be almost empty

I'd leave it here, as it's the main point of that extra file! :)